pl_bolts/callbacks/ssl_online.py

```diff
@@ -1,74 +1,96 @@
-from typing import Optional, Sequence, Tuple, Union
+from contextlib import contextmanager
+from typing import Any, Dict, Optional, Sequence, Tuple, Union
 
 import torch
 from pytorch_lightning import Callback, LightningModule, Trainer
-from torch import Tensor, device
+from pytorch_lightning.utilities import rank_zero_warn
+from torch import Tensor, nn
 from torch.nn import functional as F
 from torch.optim import Optimizer
 from torchmetrics.functional import accuracy
 
+from pl_bolts.models.self_supervised.evaluator import SSLEvaluator
+
 
 class SSLOnlineEvaluator(Callback):  # pragma: no cover
     """Attaches a MLP for fine-tuning using the standard self-supervised protocol.
 
     Example::
 
-        # your model must have 2 attributes
+        # your datamodule must have 2 attributes
+        dm = DataModule()
+        dm.num_classes = ... # the num of classes in the datamodule
+        dm.name = ... # name of the datamodule (e.g. ImageNet, STL10, CIFAR10)
+
+        # your model must have 1 attribute
         model = Model()
         model.z_dim = ... # the representation dim
-        model.num_classes = ... # the num of classes in the model
 
         online_eval = SSLOnlineEvaluator(
-            z_dim=model.z_dim,
-            num_classes=model.num_classes,
-            dataset='imagenet'
+            z_dim=model.z_dim
         )
     """
 
     def __init__(
         self,
-        dataset: str,
+        z_dim: int,
         drop_p: float = 0.2,
         hidden_dim: Optional[int] = None,
-        z_dim: int = None,
-        num_classes: int = None,
     ):
         """
         Args:
-            dataset: if stl10, need to get the labeled batch
+            z_dim: Representation dimension
             drop_p: Dropout probability
             hidden_dim: Hidden dimension for the fine-tune MLP
-            z_dim: Representation dimension
-            num_classes: Number of classes
         """
         super().__init__()
 
+        self.z_dim = z_dim
         self.hidden_dim = hidden_dim
         self.drop_p = drop_p
-        self.optimizer: Optimizer
 
-        self.z_dim = z_dim
-        self.num_classes = num_classes
-        self.dataset = dataset
+        self.optimizer: Optional[Optimizer] = None
+        self.online_evaluator: Optional[SSLEvaluator] = None
+        self.z_dim: Optional[int] = None
+        self.num_classes: Optional[int] = None
+        self.dataset: Optional[str] = None
```

Comment:
how these are set?

Reply:
They are set during `setup()`

Reply:
See [here](https://github.com/zhengyu-yang/lightning-bolts/blob/edbfca7ee9a283df2d9aee3e8705cb1c53bffe3f/pl_bolts/callbacks/ssl_online.py#L59-L61)

Reply:
well, can we still keep the init arguments and if needed set default as None?
or pls what is the reason for removing these arguments from init?

Reply:
This is more of a design choice. As this is adapted from my research code, and in my experiments, I try to remove any redundancy in configuration. 
I can add them back for backward compatibility.

Reply:
> I can add them back for backward compatibility. Do you think it is necessary?

yes, pls add it back as it does not hurt you and we do not break any user code...

Reply:
Thanks @zhengyu-yang as @Borda suggested, it would be nice to keep the signature the same for BW compat :)